samples/android/tutorial-4-opencl/jni/GLrender.cpp

```diff
@@ -0,0 +1,315 @@
+    GLuint fshader = glCreateShader(GL_FRAGMENT_SHADER);
+    text = fss;
+    glShaderSource(fshader, 1, &text, 0);
+    glCompileShader(fshader);
+    glGetShaderiv(fshader, GL_COMPILE_STATUS, &compiled);
+    if (compiled == 0) {
+        logShaderCompileError(fshader);
+        glDeleteShader(fshader);
+        fshader = 0;
+    }
+
+    LOGD("makeShaderProg: glCreateProgram");
+    GLuint program = glCreateProgram();
+    glAttachShader(program, vshader);
+    glAttachShader(program, fshader);
+    glLinkProgram(program);
+
+    if(vshader) glDeleteShader(vshader);
+    if(fshader) glDeleteShader(fshader);
+
+    return program;
+}
+
+
+static void initFBO(int width, int height)
+{
+    LOGD("initFBO(%d, %d)", width, height);
+    releaseFBO();
+
+    glGenTextures(1, &FBOtex2);
+    glBindTexture(GL_TEXTURE_2D, FBOtex2);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, 0);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+    glGenTextures(1, &FBOtex);
+    glBindTexture(GL_TEXTURE_2D, FBOtex);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, 0);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+    //int hFBO;
+    glGenFramebuffers(1, &FBO);
+    glBindFramebuffer(GL_FRAMEBUFFER, FBO);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, FBOtex, 0);
+    LOGD("initFBO status: %d", glGetError());
+
+    if (glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE)
+        LOGE("initFBO failed: %d", glCheckFramebufferStatus(GL_FRAMEBUFFER));
+
+    prog2D = makeShaderProg(vss, fss2D);
+}
+
+void drawTex(int tex, GLenum texType, GLuint fbo)
+{
+    int64_t t = getTimeMs();
+    //draw texture to FBO or to screen
+    glBindFramebuffer(GL_FRAMEBUFFER, fbo);
+    glViewport(0, 0, texWidth, texHeight);
+
+    glClear(GL_COLOR_BUFFER_BIT);
+
+    int prog = texType == GL_TEXTURE_EXTERNAL_OES ? progOES : prog2D;
+    glUseProgram(prog);
+    int vPos  = glGetAttribLocation(prog, "vPosition");
+    int vTC = glGetAttribLocation(prog, "vTexCoord");
+
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(texType, tex);
+    glUniform1i(glGetUniformLocation(prog, "sTexture"), 0);
+
+    glVertexAttribPointer(vPos, 2, GL_FLOAT, false, 4*2, vertexes);
+    glVertexAttribPointer(vTC,  2, GL_FLOAT, false, 4*2, texType == GL_TEXTURE_EXTERNAL_OES ? texCoordOES : texCoord2D);
+    glEnableVertexAttribArray(vPos);
+    glEnableVertexAttribArray(vTC);
```

Comment:
An approach without VAO is deprecated in modern OpenGL. 

> simpler

Yes. Your drawing function will not be over-complicated.

> faster

In a theory - yes. Your application is simple, so you don't notice performance improvements, but it can be seen in more complex pipeline.